x/erc20/keeper/ibc_callbacks.go

```diff
@@ -159,52 +181,59 @@ func (k Keeper) OnTimeoutPacket(ctx sdk.Context, _ channeltypes.Packet, data tra
 }
 
 // ConvertCoinToERC20FromPacket converts the IBC coin to ERC20 after refunding the sender
+// This function is only executed when IBC timeout or an Error ACK happens.
 func (k Keeper) ConvertCoinToERC20FromPacket(ctx sdk.Context, data transfertypes.FungibleTokenPacketData) error {
 	sender, err := sdk.AccAddressFromBech32(data.Sender)
 	if err != nil {
 		return err
 	}
 
-	// use a zero gas config to avoid extra costs for the relayers
-	ctx = ctx.
-		WithKVGasConfig(storetypes.GasConfig{}).
-		WithTransientKVGasConfig(storetypes.GasConfig{})
-
-	// assume that all module accounts on Evmos need to have their tokens in the
-	// IBC representation as opposed to ERC20
-	senderAcc := k.accountKeeper.GetAccount(ctx, sender)
-	if types.IsModuleAccount(senderAcc) {
+	pairID := k.GetTokenPairID(ctx, data.Denom)
+	pair, found := k.GetTokenPair(ctx, pairID)
+	if !found {
+		// no-op, token pair is not registered
 		return nil
 	}
 
 	coin := ibc.GetSentCoin(data.Denom, data.Amount)
 
-	// check if the coin is a native staking token
-	bondDenom := k.stakingKeeper.BondDenom(ctx)
-	if coin.Denom == bondDenom {
-		// no-op, received coin is the staking denomination
-		return nil
-	}
+	switch {
 
-	params := k.GetParams(ctx)
-	if !params.EnableErc20 || !k.IsDenomRegistered(ctx, coin.Denom) {
-		// no-op, ERC20s are disabled or the denom is not registered
+	// Case 1. if pair is native coin -> no-op
+	case pair.IsNativeCoin():
+		// no-op, received coin is a  native coin
 		return nil
-	}
-
-	msg := types.NewMsgConvertCoin(coin, common.BytesToAddress(sender), sender)
-
-	// NOTE: we don't use ValidateBasic the msg since we've already validated the
-	// fields from the packet data
 
-	// convert Coin to ERC20
-	if _, err = k.ConvertCoin(sdk.WrapSDKContext(ctx), msg); err != nil {
-		return err
+	// Case 2. if pair is native ERC20 -> unescrow
+	case pair.IsNativeERC20():
+		// use a zero gas config to avoid extra costs for the relayers
+		ctx = ctx.
+			WithKVGasConfig(storetypes.GasConfig{}).
+			WithTransientKVGasConfig(storetypes.GasConfig{})
+
+		params := k.GetParams(ctx)
+		if !params.EnableErc20 || !k.IsDenomRegistered(ctx, coin.Denom) {
+			// no-op, ERC20s are disabled or the denom is not registered
+			return nil
+		}
+
+		// assume that all module accounts on Evmos need to have their tokens in the
+		// IBC representation as opposed to ERC20
+		senderAcc := k.accountKeeper.GetAccount(ctx, sender)
+		if types.IsModuleAccount(senderAcc) {
+			return nil
+		}
+
+		// Convert from Coin to ERC20
+		if err := k.ConvertCoinNativeERC20(ctx, pair, coin.Amount, common.BytesToAddress(sender), sender); err != nil {
+			// We want to record only the failed attempt to reconvert the coins during IBC.
+			defer func() {
+				telemetry.IncrCounter(1, types.ModuleName, "ibc", "error", "total")
+			}()
```

Comment:
Nit, no need the defer here
```suggestion
			
			telemetry.IncrCounter(1, types.ModuleName, "ibc", "error", "total")
			
```